### PR TITLE
Anchor Design Spec

### PR DESCRIPTION
**Pull request by AdityaSripal**
Initial design for Anchor Chain

Initial thoughts on how sidechain would interact with checkpointing service.

### CODE REVIEW

File: PlasmaAnchor/anchor.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+Let the current objectively finalized block be `B` and the currently objectively finalized deposit be `N`.
+
+### Msgs
+
+Msgs in anchor are submitted by each sidechain validator once per block
+
+```go
+// Implement sdk.Msg interface
+type AnchorMsg struct {
```
Comment (AdityaSripal) [inline PlasmaAnchor/anchor.md:25]: what happens if this gets too big?

File: PlasmaAnchor/anchor.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+Since there are a limited number of sidechain validators and the lists are limited to being at most the length of new data that has been added to rootchain since last objectively finalized checkpoint, this isn't an infeasibly expensive calculation to happen in `endBlocker`.
+
+### Subjective Finality
+
+In order to preserve the notion of subjective finality, validators define a parameter in their anchor node's `config.toml` file called `subjective-finality`. This is how many blocks the validator believes need to be added after a piece of data before that data is presumed finalized by the validator. The Anchor node will then poll the rootchain node continuously and pull all relevant data that has been subjectively finalized according to the above parameter but not objectively finalized by Anchor and construct an `AnchorMsg` submitting that data to Anchor chain.
+
+To tune their `subjective-finality` parameter, validators simply stop their anchor node, update their `config.toml` with their preferred `subjective-finality` bound, and restart the anchor node.
+
+### Catastrophic Failure
```
Comment (AdityaSripal) [inline PlasmaAnchor/anchor.md:68]: Future work should look to address this, would be irrelevant for Plasma but worth doing in general case

File: PlasmaAnchor/anchor.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+Let the current objectively finalized block be `B` and the currently objectively finalized deposit be `N`.
+
+### Msgs
+
+Msgs in anchor are submitted by each sidechain validator once per block
+
+```go
+// Implement sdk.Msg interface
+type AnchorMsg struct {
```
Comment (AdityaSripal) [inline PlasmaAnchor/anchor.md:25]: Should have better name for Msg

File: PlasmaAnchor/anchor.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+Having a single node act as an oracle for the sidechain would cause a central point of failure. Including the anchor mechanism in the sidechain consensus state introduces a lot of complexity and is not very modular. Anchor is a seperate process that will run Tendermint consensus to determine the latest rootchain state that a quorum of validators deem finalized. This will be used to communicate to the sidechain the latest finalized rootchain state to process deposits and exits.
+
+The sidechain must track `Deposits` in the rootchain so that users can spend finalized deposits on the sidechain. The sidechain should also track `Exits` in order to delete positions that are no longer spendable (space-saving) and prevent users from spamming the chain by spending exitted positions.
```
Comment (colin-axner) [inline PlasmaAnchor/anchor.md]: "exitted" is misspelled. Should be 'exited'

File: PlasmaAnchor/anchor.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+Rootchain Blocks: `Map{Number -> BlockHash}`
+
+Deposits: `Map{Number -> Deposit}`
+
+where `Deposit` is:
+
+```go
+type Deposit struct {
+    Owner  ethcmn.Address
```
Comment (colin-axner) [inline PlasmaAnchor/anchor.md]: Denomination should be included

Comment (AdityaSripal) [reply PlasmaAnchor/anchor.md]: That is what amount is? Do you mean rename to `Denomination` or do you mean specify type of Deposit

i.e. Denomination is string `Eth`, `OMG`, whatever

Comment (colin-axner) [reply PlasmaAnchor/anchor.md]: Amount of the deposit needs to be included

Comment (AdityaSripal) [reply PlasmaAnchor/anchor.md]: It already is. Checkout the line right below, Line 22

Comment (colin-axner) [reply PlasmaAnchor/anchor.md]: Whoops, read that as the deposit nonce for some reason

File: PlasmaAnchor/anchor.md (hunk 1 above)
Comment (AdityaSripal) [reply PlasmaAnchor/anchor.md:25]: Now capped

**Review by colin-axner — APPROVED** [review]

File: PlasmaAnchor/anchor.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+Msgs in anchor are submitted by each sidechain validator once per block
+
+```go
+// Implement sdk.Msg interface
+type AnchorMsg struct {
+    // Msg contains all subjectively finalized data 
+    // since last objectively finalized data
+    RootchainBlockHash [][]byte
+    Deposits           []Deposit
```
Comment (hamdiallam) [inline PlasmaAnchor/anchor.md]: Why must deposits/exits be passed around as messages in the network?

If the blockhash is already known, can't this information be retrieved locally from the geth node? I was thinking each client would temporarily collect deposits/exits locally per block

File: PlasmaAnchor/anchor.md [hunk 7]
```diff
@@ -0,0 +1,117 @@
+## Anchor: Finality Gadget on Tendermint
+
+### Motivation
```
Comment (hamdiallam) [inline PlasmaAnchor/anchor.md:3]: I think it would be nice to state at the beginning that this mainly needed for a validator set > 1

**Comment by hamdiallam** [comment]
Optimization suggestion:

We can cut down on the messages sent around per block in the validator set. Rather than each validator node sending the entire list of headers seen past the last finalized header, each validator should maintain a transient view of headers seen by its peers, discarding headers that are come before the most recent finalized point.  

This allows validators to only send around headers that they have most currently seen and not repeat old headers.

**Review by hamdiallam — APPROVED** [review]

**Pull request merged**